src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,373 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<
+            TInvocationSyntax,
+            TExpressionSyntax,
+            TMethodDeclarationSyntax,
+            TStatementSyntax>
+        : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationSyntax : TExpressionSyntax
+        where TMethodDeclarationSyntax : SyntaxNode
+        where TStatementSyntax : SyntaxNode
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFactsService;
+
+        protected abstract TExpressionSyntax? GetInlineExpression(TMethodDeclarationSyntax calleeMethodDeclarationSyntaxNode);
+        protected abstract SyntaxNode? GetEnclosingMethodLikeNode(SyntaxNode syntaxNode);
+        protected abstract SyntaxNode GenerateTypeSyntax(ITypeSymbol symbol, bool allowVar);
+        protected abstract TExpressionSyntax Parenthesize(TExpressionSyntax expressionNode);
+        protected abstract bool TryGetInlineSyntaxNodeAndReplacementNodeForDelegate(
+            TInvocationSyntax calleeInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            TExpressionSyntax inlineExpressionNode,
+            TStatementSyntax statementContainsCallee,
+            SyntaxGenerator syntaxGenerator,
+            out SyntaxNode? inlineSyntaxNode,
+            out SyntaxNode? syntaxNodeToReplace);
+
+        /// <summary>
+        /// Check if <paramref name="expressionNode"/> could be used as an Expression in ExpressionStatement
+        /// </summary>
+        protected abstract bool IsValidExpressionUnderStatementExpression(TExpressionSyntax expressionNode);
+
+        /// <summary>
+        /// A preferred name used to generated a declaration when the
+        /// inline method has a return value but is not assigned to a variable.
```

Comment:
```suggestion
        /// inline method has a return value but is not assigned to a variable.  Importantly, it should not be localized
        /// as we need to ensure we can generate a legal C#/VB identifier from it.
```

Reply:
This part is removed becasue in some other comment there is a better way to handle type cast.